### PR TITLE
Change timezone to only accept valid timezones

### LINKED ISSUE #282: [PE-D] Unable to input non-integer GMT timezones
Certain parts of the world are on non-integer GMT timezones. For example, Indian is on GMT+5.5. This means that I cannot accurately track the timezones of any clients living in India.

-------------
Labels: `severity.Medium` `type.FeatureFlaw`
original: ianyong/ped#3

### PR DESCRIPTION
**Pull request by tankangliang**
## Description

`Timezone` class now references a set of valid timezones for validation. The format for parameters is now `GMT+HH:MM` and the list of valid timezones can be found [here](https://www.timeanddate.com/time/current-number-time-zones.html).

Fixes #282

## Testing

As per UserGuide:

`client add n/Katya p/98123456 e/katya@yahoo.com a/Vladivostok, Nevelskogo, bld. 15, appt. 256 c/RU tz/GMT+03:00 ce/22-12-2020`
`client edit 3 c/JP tz/GMT+07:00`

Also manual testing with `client view 1` and changing the timezones to verify they are of the correct timing.

### CODE REVIEW

File: docs/UserGuide.md [hunk 1]
```diff
@@ -535,7 +535,7 @@ Parameter | Prefix | Constraints, Examples
 **EMAIL** | `e/` | Emails should be of the format local-part@domain. <br> e.g. `e/katya@yahoo.com`
 **ADDRESS** |`a/` | Addresses can take any values, and it should not be blank. <br> e.g. `Vladivostok, Nevelskogo, bld. 15, appt. 256`
 **COUNTRY_CODE** | `c/` | A 2-letter country code that follows the ISO3166 specification <br> This [finding tool](https://www.countrycode.org/) can be used. <br> e.g. `c/SG` (Singapore) 
-**TIMEZONE** | `tz/` | Timezone should be given in offsets relative to [Greenwich Mean Time](https://en.wikipedia.org/wiki/Greenwich_Mean_Time). <br> e.g. `tz/GMT+8`
+**TIMEZONE** | `tz/` | Timezone should be given in offsets relative to [Greenwich Mean Time](https://en.wikipedia.org/wiki/Greenwich_Mean_Time) in the format `GMT+HH:MM` where `HH` refers to the offset in hours and `MM` refers to the offset in minutes. The full list of valid timezones can be found [here](https://www.timeanddate.com/time/current-number-time-zones.html) <br> e.g. `tz/GMT+08:00
```
Comment (qwoprocks) [inline docs/UserGuide.md]: the link leads to a table with UTC timezones instead

File: src/main/java/seedu/address/model/client/Timezone.java [hunk 2]
```diff
@@ -17,21 +22,31 @@
...
-    public static final int LARGEST_POSITIVE_OFFSET = 14;
 
-    public static final String MESSAGE_CONSTRAINTS =
-            "Timezone should be in the form \"" + GMT_STRING + "+X\" or \"" + GMT_STRING
-            + "-X\" where X is a number.\n" + "Largest offset is " + GMT_STRING
-            + "+" + LARGEST_POSITIVE_OFFSET + " and smallest offset is " + GMT_STRING
-            + "-" + SMALLEST_NEGATIVE_OFFSET;
+    public static final Set<String> VALID_TIMEZONES = new HashSet<>();
+
+    static {
```
Comment (qwoprocks) [inline src/main/java/seedu/address/model/client/Timezone.java:26]: I think we should standardize (forgot which other file was it) and just remove the static, and add a comment above saying this is some special static init block

File: src/main/java/seedu/address/model/client/Timezone.java [hunk 3]
```diff
@@ -17,21 +22,31 @@
...
+        Set<String> allZones = ZoneId.getAvailableZoneIds();
+        List<String> zoneList = new ArrayList<>(allZones);
+        for (String s : zoneList) {
+            ZoneOffset offset = LocalDateTime.now().atZone(ZoneId.of(s)).getOffset();
+            String out = String.format("%s", offset);
+            VALID_TIMEZONES.add(out);
+        }
```
Comment (qwoprocks) [inline src/main/java/seedu/address/model/client/Timezone.java]: Is there a need to convert to arraylist? Cannot just iterate through the set?

File: src/main/java/seedu/address/model/client/Timezone.java [hunk 4]
```diff
@@ -17,21 +22,31 @@
...
+        VALID_TIMEZONES.remove("Z");
+        VALID_TIMEZONES.add("+00:00");
```
Comment (qwoprocks) [inline src/main/java/seedu/address/model/client/Timezone.java:35]: Should probably add a comment stating what this does.

File: src/main/java/seedu/address/model/client/Timezone.java [hunk 5]
```diff
@@ -17,21 +22,31 @@
...
+    }
+
+    public static final String MESSAGE_CONSTRAINTS = "Timezone should be in the form \"" + GMT_STRING + "+HH:MM\" or \""
+            + GMT_STRING + "-HH:MM\" where HH is the offset in hours and MM is the offset in minutes. The full list of "
+            + "valid timezones can be found at https://www.timeanddate.com/time/current-number-time-zones.html";
```
Comment (qwoprocks) [inline src/main/java/seedu/address/model/client/Timezone.java:40]: same comment as in UG about the link. Is there any better links?

File: src/main/java/seedu/address/model/client/Timezone.java [hunk 6]
```diff
@@ -76,35 +82,36 @@ public static boolean isValidTimezone(String test) {
...
 
     @Override
     public String toString() {
-        // Negative offset values will automatically include "-"
-        String sign = offsetValue >= 0 ? "+" : "";
-        return GMT_STRING + sign + offsetValue;
+        String zoneOffsetIdString = zoneOffsetId.toString();
+        if (zoneOffsetIdString.equals(GMT_STRING)) {
+            zoneOffsetIdString = GMT_STRING + "+00:00";
+        }
```
Comment (qwoprocks) [inline src/main/java/seedu/address/model/client/Timezone.java]: Is this some special case? Should add a comment if that's the case

File: src/test/java/seedu/address/model/client/TimezoneTest.java [hunk 7]
```diff
@@ -32,65 +32,106 @@ public void isValidTimezone() {
...
+            int hoursOffset = Integer.parseInt(sign + hoursOffsetString);
+            int minutesOffset = Integer.parseInt(sign + minutesOffsetString);
+
             long totalMilliseconds = System.currentTimeMillis();
             long totalSeconds = totalMilliseconds / 1000;
             long totalMinutes = totalSeconds / 60;
-            long totalHours = totalMinutes / 60;
-            long currentHour = ((totalHours + timezoneOffset) % 24);
+            long totalHours = (totalMinutes + minutesOffset) / 60;
+            long currentHour = ((totalHours + hoursOffset) % 24);
```
Comment (qwoprocks) [inline src/test/java/seedu/address/model/client/TimezoneTest.java]:
```suggestion
            long currentHour = (totalHours + hoursOffset) % 24;
```

File: src/test/java/seedu/address/model/client/TimezoneTest.java [hunk 8]
```diff
@@ -32,65 +32,106 @@ public void isValidTimezone() {
...
-        assertFalse(Timezone.isValidTimezone("UTC+8")); // does not start with GMT
-        assertFalse(Timezone.isValidTimezone("GM+8")); // misspelled GMT
-        assertFalse(Timezone.isValidTimezone("GMT+15")); // out of range (positive)
-        assertFalse(Timezone.isValidTimezone("GMT-13")); // out of range (negative)
+        assertFalse(Timezone.isValidTimezone("UTC+00:00")); // does not start with GMT
+        assertFalse(Timezone.isValidTimezone("GM+00:00")); // misspelled GMT
+        assertFalse(Timezone.isValidTimezone("GMT+15:00")); // out of range (positive)
+        assertFalse(Timezone.isValidTimezone("GMT-13:00")); // out of range (negative)
+        assertFalse(Timezone.isValidTimezone("GMT+13:30")); // unrecognised timezone
         assertFalse(Timezone.isValidTimezone("GMT+654657987654456")); // large number
```
Comment (qwoprocks) [inline src/test/java/seedu/address/model/client/TimezoneTest.java]: Should add a few more negative test cases like "GMT" and "GMT+" and "GMT-", and maybe "gmt+08:00", and "GMT+8" and "GMT+08", and "GMT+8:00", and "GMT+08:0"

File: src/test/java/seedu/address/logic/commands/ClientAddCommandIntegrationTest.java [hunk 9]
```diff
@@ -29,6 +29,7 @@ public void setUp() {
     public void execute_newClient_success() {
         Client validClient = new ClientBuilder().build();
 
+
```
Comment (qwoprocks) [inline src/test/java/seedu/address/logic/commands/ClientAddCommandIntegrationTest.java]: extra newline

**Review by qwoprocks — CHANGES_REQUESTED** [review]

File: docs/UserGuide.md (hunk 1 above)
Comment (rtshkmr) [reply docs/UserGuide.md]: can this be used instead? https://greenwichmeantime.com/time-gadgets/time-zone-converter/

File: src/main/java/seedu/address/model/client/Timezone.java (hunk 5 above)
Comment (rtshkmr) [reply src/main/java/seedu/address/model/client/Timezone.java:40]: can this be used instead? https://greenwichmeantime.com/time-gadgets/time-zone-converter/

File: src/main/java/seedu/address/model/client/Timezone.java (hunk 2 above)
Comment (tankangliang) [reply src/main/java/seedu/address/model/client/Timezone.java:26]: I only found the one in `TbmManager.java` but that one doesn't seem to be for class-level variables

Comment (qwoprocks) [reply src/main/java/seedu/address/model/client/Timezone.java:26]: Oh right ok

File: src/main/java/seedu/address/model/client/Timezone.java (hunk 4 above)
Comment (qwoprocks) [reply src/main/java/seedu/address/model/client/Timezone.java:35]: bump

File: src/main/java/seedu/address/model/client/Timezone.java [hunk 10]
```diff
@@ -41,11 +53,9 @@
     public Timezone(String timezone) {
         requireNonNull(timezone);
         checkArgument(isValidTimezone(timezone), MESSAGE_CONSTRAINTS);
-        String offsetString = timezone.substring(GMT_STRING.length());
-        int offsetValue = Integer.parseInt(offsetString);
+        assert (VALID_TIMEZONES.contains(timezone.substring(UTC_STRING.length())));
```
Comment (qwoprocks) [inline src/main/java/seedu/address/model/client/Timezone.java]:
```suggestion
        assert VALID_TIMEZONES.contains(timezone.substring(UTC_STRING.length()));
```
btw what does this assert do? Isit rly needed lmao

**Review by qwoprocks — APPROVED** [review]
Small comments, but LGTM

**Pull request merged**